File: backend/src/services/strategy-execution/strategy-executor.ts
/**
 * Strategy Executor Service
 * Orchestrates strategy execution in both backtest and live modes
 * Handles multi-tenant execution with encrypted API key management
 */

import { spawn } from 'child_process';
import * as path from 'path';
import * as fs from 'fs/promises';
import logger from '../../config/logger';
import { strategyEnvironmentManager } from '../strategy-environment-manager';
import { decrypt } from '../../utils/encryption';
import { prisma } from '../../config/database';

export interface StrategySettings {
  strategy_id: string;
  pair: string;
  capital: number;
  leverage: number;
  risk_per_trade: number;
  resolution?: string;
  commission_rate?: number;
  gst_rate?: number;
  sl_rate?: number;
  tp_rate?: number;
  start_date?: string;
  end_date?: string;
}

export interface Subscriber {
  user_id: string;
  api_key: string;      // Decrypted in memory only
  api_secret: string;   // Decrypted in memory only
  capital: number;
  leverage: number;
  risk_per_trade: number;
}

export interface BacktestResult {
  success: boolean;
  mode: 'backtest';
  metrics?: any;
  trades?: any[];
  total_trades?: number;
  error?: string;
  traceback?: string;
}

export interface LiveExecutionResult {
  success: boolean;
  mode: 'live';
  result?: any;
  subscribers_count?: number;
  error?: string;
  traceback?: string;
}

export type ExecutionResult = BacktestResult | LiveExecutionResult;

export class StrategyExecutor {
  /**
   * Get the strategy file path for a given strategy ID
   */
  private async getStrategyFilePath(strategyId: string): Promise<string> {
    const strategiesDir = path.join(__dirname, '../../../strategies');
    const strategyDir = path.join(strategiesDir, strategyId);

    // Find the .py file in the strategy directory
    const files = await fs.readdir(strategyDir);
    const pyFile = files.find(f => f.endsWith('.py'));

    if (!pyFile) {
      throw new Error(`No Python file found in strategy directory: ${strategyDir}`);
    }

    return path.join(strategyDir, pyFile);
  }

  /**
   * Execute strategy in backtest mode
   */
  async executeBacktest(
    strategyId: string,
    settings: StrategySettings
  ): Promise<BacktestResult> {
    logger.info(`Executing backtest for strategy ${strategyId}`);

    try {
      // Ensure environment exists
      const envExists = await strategyEnvironmentManager.environmentExists(strategyId);
      if (!envExists) {
        logger.info(`Environment not found for strategy ${strategyId}, creating...`);
        await strategyEnvironmentManager.createEnvironment(strategyId);
      }

      // Get environment info
      const envInfo = await strategyEnvironmentManager.getEnvironmentInfo(strategyId);
      if (envInfo.status !== 'ready') {
        throw new Error(`Environment is not ready for strategy ${strategyId}: ${envInfo.status}`);
      }

      // Get strategy file path
      const strategyFilePath = await this.getStrategyFilePath(strategyId);

      // Prepare input for Python executor
      const input = {
        mode: 'backtest',
        strategy_file: strategyFilePath,
        settings: {
          ...settings,
          strategy_id: strategyId,
        },
      };

      // Execute in isolated environment
      const result = await this.executePythonScript(
        envInfo.pythonPath,
        path.join(__dirname, '../../../python/strategy_executor.py'),
        input
      );

      // Parse result
      const parsedResult = JSON.parse(result.stdout);

      // Log result
      if (parsedResult.success) {
        logger.info(`Backtest completed successfully for strategy ${strategyId}`);
        logger.info(`Total trades: ${parsedResult.total_trades}`);
      } else {
        logger.error(`Backtest failed for strategy ${strategyId}: ${parsedResult.error}`);
      }

      return parsedResult as BacktestResult;

    } catch (error) {
      logger.error(`Failed to execute backtest for strategy ${strategyId}:`, error);
      return {
        success: false,
        mode: 'backtest',
        error: error instanceof Error ? error.message : 'Unknown error',
      };
    }
  }

  /**
   * Fetch active subscribers with decrypted credentials
   */
  private async fetchActiveSubscribers(strategyId: string): Promise<Subscriber[]> {
    logger.info(`Fetching active subscribers for strategy ${strategyId}`);

    try {
      // Fetch subscriptions with broker credentials
      const subscriptions = await prisma.strategySubscription.findMany({
        where: {
          strategyId,
          isActive: true,
          isPaused: false,
        },
        include: {
          brokerCredential: {
            select: {
              apiKey: true,     // Encrypted in DB
              apiSecret: true,  // Encrypted in DB
            },
          },
        },
      });

      // Decrypt credentials in memory
      const subscribers: Subscriber[] = subscriptions.map((sub) => ({
        user_id: sub.userId,
        api_key: decrypt(sub.brokerCredential.apiKey),      // Decrypt in memory
        api_secret: decrypt(sub.brokerCredential.apiSecret), // Decrypt in memory
        capital: sub.capital,
        leverage: sub.leverage,
        risk_per_trade: sub.riskPerTrade,
      }));

      logger.info(`Found ${subscribers.length} active subscribers for strategy ${strategyId}`);

      return subscribers;

    } catch (error) {
      logger.error(`Failed to fetch subscribers for strategy ${strategyId}:`, error);
      throw error;
    }
  }

  /**
   * Execute strategy in live mode
   */
  async executeLive(
    strategyId: string,
    settings: StrategySettings
  ): Promise<LiveExecutionResult> {
    logger.info(`Executing live trading for strategy ${strategyId}`);

    try {
      // Fetch active subscribers with decrypted credentials
      const subscribers = await this.fetchActiveSubscribers(strategyId);

      if (subscribers.length === 0) {
        logger.warn(`No active subscribers found for strategy ${strategyId}`);
        return {
          success: false,
          mode: 'live',
          error: 'No active subscribers found',
        };
      }

      // Ensure environment exists
      const envExists = await strategyEnvironmentManager.environmentExists(strategyId);
      if (!envExists) {
        throw new Error(`Environment not found for strategy ${strategyId}. Strategy must be validated first.`);
      }

      // Get environment info
      const envInfo = await strategyEnvironmentManager.getEnvironmentInfo(strategyId);
      if (envInfo.status !== 'ready') {
        throw new Error(`Environment is not ready for strategy ${strategyId}: ${envInfo.status}`);
      }

      // Get strategy file path
      const strategyFilePath = await this.getStrategyFilePath(strategyId);

      // Prepare input for Python executor
      const input = {
        mode: 'live',
        strategy_file: strategyFilePath,
        settings: {
          ...settings,
          strategy_id: strategyId,
        },
        subscribers,
      };

      // Execute in isolated environment
      const result = await this.executePythonScript(
        envInfo.pythonPath,
        path.join(__dirname, '../../../python/strategy_executor.py'),
        input
      );

      // Parse result
      const parsedResult = JSON.parse(result.stdout);

      // Log result
      if (parsedResult.success) {
        logger.info(`Live execution completed successfully for strategy ${strategyId}`);
        logger.info(`Executed for ${parsedResult.subscribers_count} subscribers`);
      } else {
        logger.error(`Live execution failed for strategy ${strategyId}: ${parsedResult.error}`);
      }

      // Update trade statistics in database
      if (parsedResult.success) {
        await this.updateTradeStatistics(strategyId, subscribers.length);
      }

      return parsedResult as LiveExecutionResult;

    } catch (error) {
      logger.error(`Failed to execute live trading for strategy ${strategyId}:`, error);
      return {
        success: false,
        mode: 'live',
        error: error instanceof Error ? error.message : 'Unknown error',
      };
    }
  }

  /**
   * Execute Python script with JSON input/output
   */
  private executePythonScript(
    pythonPath: string,
    scriptPath: string,
    input: any
  ): Promise<{ stdout: string; stderr: string }> {
    return new Promise((resolve, reject) => {
      const process = spawn(pythonPath, [scriptPath], {
        stdio: ['pipe', 'pipe', 'pipe'],
      });

      let stdout = '';
      let stderr = '';

      // Send input as JSON to stdin
      process.stdin.write(JSON.stringify(input));
      process.stdin.end();

      process.stdout.on('data', (data) => {
        stdout += data.toString();
      });

      process.stderr.on('data', (data) => {
        stderr += data.toString();
      });

      process.on('close', (code) => {
        if (code === 0) {
          resolve({ stdout, stderr });
        } else {
          reject(new Error(`Process exited with code ${code}\nStderr: ${stderr}\nStdout: ${stdout}`));
        }
      });

      process.on('error', (error) => {
        reject(error);
      });
    });
  }

  /**
   * Update trade statistics after successful execution
   */
  private async updateTradeStatistics(strategyId: string, subscriberCount: number): Promise<void> {
    try {
      // Update strategy statistics
      await prisma.strategy.update({
        where: { id: strategyId },
        data: {
          lastExecutedAt: new Date(),
          // You can add more statistics here as needed
        },
      });

      logger.info(`Updated trade statistics for strategy ${strategyId}`);
    } catch (error) {
      logger.error(`Failed to update trade statistics for strategy ${strategyId}:`, error);
      // Don't throw - this is non-critical
    }
  }

  /**
   * Get strategy resolution from STRATEGY_CONFIG in the Python file
   */
  async getStrategyResolution(strategyId: string): Promise<string> {
    try {
      const strategyFilePath = await this.getStrategyFilePath(strategyId);
      const content = await fs.readFile(strategyFilePath, 'utf-8');

      // Simple regex to find STRATEGY_CONFIG['resolution'] or STRATEGY_CONFIG["resolution"]
      const resolutionMatch = content.match(/STRATEGY_CONFIG\s*=\s*{[^}]*['"]resolution['"]\s*:\s*['"](\w+)['"]/);

      if (resolutionMatch && resolutionMatch[1]) {
        const resolution = resolutionMatch[1];
        logger.info(`Found resolution '${resolution}' for strategy ${strategyId}`);
        return resolution;
      }

      // Default to 5 minutes if not found
      logger.warn(`No resolution found in STRATEGY_CONFIG for strategy ${strategyId}, defaulting to '5m'`);
      return '5m';

    } catch (error) {
      logger.error(`Failed to get resolution for strategy ${strategyId}:`, error);
      return '5m'; // Default fallback
    }
  }
}

// Export singleton instance
export const strategyExecutor = new StrategyExecutor();

File: backend/python/strategy_executor.py
"""
Strategy Executor
Unified entry point for executing trading strategies in both backtest and live modes.
Reads JSON input from stdin and outputs JSON results to stdout.
"""

import sys
import json
import os
import importlib.util
from typing import Dict, Any, List, Optional
from pathlib import Path
import traceback
from concurrent.futures import ThreadPoolExecutor, as_completed


def load_strategy_module(strategy_file_path: str):
    """
    Dynamically load a strategy Python file as a module

    Args:
        strategy_file_path: Absolute path to the strategy .py file

    Returns:
        Loaded module object
    """
    if not os.path.exists(strategy_file_path):
        raise FileNotFoundError(f"Strategy file not found: {strategy_file_path}")

    # Create module spec and load
    spec = importlib.util.spec_from_file_location("strategy_module", strategy_file_path)
    if spec is None or spec.loader is None:
        raise ImportError(f"Failed to create module spec for {strategy_file_path}")

    module = importlib.util.module_from_spec(spec)
    sys.modules["strategy_module"] = module
    spec.loader.exec_module(module)

    return module


def normalize_metrics(raw_metrics: Dict[str, Any]) -> Dict[str, Any]:
    """
    Normalize metrics from strategy's format to backend's expected format.

    Converts from:
        {'Win Rate (%)': '56.23%', 'Profit Factor': '2.45', ...}
    To:
        {'winRate': 56.23, 'profitFactor': 2.45, ...}

    Args:
        raw_metrics: Metrics dictionary from strategy's evaluate_backtest_metrics

    Returns:
        Normalized metrics dictionary with camelCase keys and numeric values
    """
    import re

    def parse_percentage(value: str) -> float:
        """Parse percentage string like '56.23%' to float 56.23"""
        if isinstance(value, (int, float)):
            return float(value)
        if isinstance(value, str):
            # Remove % sign, $ sign, and commas
            clean = value.replace('%', '').replace('$', '').replace(',', '').strip()
            try:
                return float(clean)
            except:
                return 0.0
        return 0.0

    # Map user's keys to backend's expected keys
    normalized = {}

    for key, value in raw_metrics.items():
        lower_key = key.lower()

        if 'win rate' in lower_key or 'winrate' in lower_key:
            normalized['winRate'] = parse_percentage(value)
        elif 'total return' in lower_key or 'roi' in lower_key:
            normalized['roi'] = parse_percentage(value)
        elif 'max drawdown' in lower_key and '%' in str(value):
            normalized['maxDrawdown'] = parse_percentage(value)
        elif 'profit factor' in lower_key:
            normalized['profitFactor'] = parse_percentage(value)
        elif 'net pnl' in lower_key:
            normalized['netPnl'] = parse_percentage(value)
        elif 'expectancy' in lower_key:
            normalized['expectancy'] = parse_percentage(value)

    return normalized


def setup_logging(strategy_dir: str, strategy_id: str, mode: str):
    """
    Setup logging using the CsvHandler from the strategy module

    Args:
        strategy_dir: Directory where strategy files are located
        strategy_id: ID of the strategy
        mode: Execution mode (backtest or live)
    """
    # Import will be done in execute_strategy after module is loaded
    pass


def execute_backtest(module, settings: Dict[str, Any]) -> Dict[str, Any]:
    """
    Execute strategy in backtest mode with progress reporting

    Args:
        module: Loaded strategy module
        settings: Strategy settings including pair, capital, leverage, etc.

    Returns:
        Backtest results as dictionary
    """
    import sys
    import time

    try:
        # Get the Backtester class from the module
        Backtester = getattr(module, 'Backtester')

        # Create backtester instance
        backtester = Backtester(settings)

        # Extract parameters from settings
        pair = settings.get('pair')
        start_date = settings.get('start_date')
        end_date = settings.get('end_date')
        resolution = settings.get('resolution', '5m')
        # Strip 'm' from resolution if present (e.g., "5m" -> "5")
        resolution = resolution.rstrip('m') if isinstance(resolution, str) else str(resolution)
        initial_capital = settings.get('capital', 10000)
        leverage = settings.get('leverage', 1)
        commission_rate = settings.get('commission_rate', 0.0005)
        gst_rate = settings.get('gst_rate', 0.18)
        sl_rate = settings.get('sl_rate', 0.02)
        tp_rate = settings.get('tp_rate', 0.04)

        # Report progress: Fetching data
        print(json.dumps({
            "type": "progress",
            "stage": "fetching_data",
            "progress": 0.1,
            "message": f"Fetching historical data for {pair}..."
        }), file=sys.stderr, flush=True)

        fetch_start = time.time()

        # Fetch historical data
        df = Backtester.fetch_coindcx_data(pair, start_date, end_date, resolution)

        if df is None or df.empty:
            return {
                "success": False,
                "error": "Failed to fetch historical data or data is empty",
                "mode": "backtest"
            }

        fetch_duration = time.time() - fetch_start
        total_candles = len(df)

        # Report progress: Data fetched
        print(json.dumps({
            "type": "progress",
            "stage": "data_fetched",
            "progress": 0.3,
            "message": f"Fetched {total_candles:,} candles in {fetch_duration:.1f}s"
        }), file=sys.stderr, flush=True)

        # Report progress: Generating signals
        print(json.dumps({
            "type": "progress",
            "stage": "generating_signals",
            "progress": 0.35,
            "message": "Generating trading signals..."
        }), file=sys.stderr, flush=True)

        # Generate signals using Trader.generate_signals (Backtester inherits from Trader)
        # Use the backtester instance to call both generate_signals AND execute_trades
        df = backtester.generate_signals(df, settings)

        # Reset DataFrame index to ensure clean sequential integer indices
        # This prevents index corruption issues with large datasets after merge_asof/dropna operations
        if df is not None and not df.empty:
            df = df.reset_index(drop=True)

        if df is None or df.empty:
            return {
                "success": False,
                "error": "Signal generation failed or returned empty dataframe",
                "mode": "backtest"
            }

        # Report progress: Running backtest
        print(json.dumps({
            "type": "progress",
            "stage": "running_backtest",
            "progress": 0.4,
            "message": f"Executing backtest on {total_candles:,} candles...",
            "total_candles": total_candles
        }), file=sys.stderr, flush=True)

        backtest_start = time.time()

        # Execute trades (use instance method, not static method)
        trades_df = backtester.execute_trades(
            df=df,
            initial_capital=initial_capital,
            leverage=leverage,
            commission_rate=commission_rate,
            gst_rate=gst_rate,
            sl_rate=sl_rate,
            tp_rate=tp_rate
        )

        backtest_duration = time.time() - backtest_start

        if trades_df is None or trades_df.empty:
            return {
                "success": False,
                "error": "No trades generated during backtest",
                "mode": "backtest"
            }

        # Report progress: Calculating metrics
        print(json.dumps({
            "type": "progress",
            "stage": "calculating_metrics",
            "progress": 0.9,
            "message": f"Calculating performance metrics...",
            "backtest_duration": round(backtest_duration, 1)
        }), file=sys.stderr, flush=True)

        # Evaluate metrics
        raw_metrics = Backtester.evaluate_backtest_metrics(trades_df, initial_capital)
        # Normalize metrics to backend's expected format (camelCase keys, numeric values)
        metrics = normalize_metrics(raw_metrics)

        # Convert trades_df to dictionary for JSON serialization
        trades_list = trades_df.to_dict('records') if not trades_df.empty else []

        total_duration = fetch_duration + backtest_duration

        # Report progress: Complete
        print(json.dumps({
            "type": "progress",
            "stage": "complete",
            "progress": 1.0,
            "message": f"Backtest complete! {len(trades_list)} trades executed in {total_duration:.1f}s"
        }), file=sys.stderr, flush=True)

        return {
            "success": True,
            "mode": "backtest",
            "metrics": metrics,
            "trades": trades_list,
            "total_trades": len(trades_list),
            "stats": {
                "total_candles": total_candles,
                "fetch_duration": round(fetch_duration, 1),
                "backtest_duration": round(backtest_duration, 1),
                "total_duration": round(total_duration, 1)
            }
        }

    except Exception as e:
        # Report error
        print(json.dumps({
            "type": "error",
            "stage": "error",
            "message": str(e)
        }), file=sys.stderr, flush=True)

        return {
            "success": False,
            "error": str(e),
            "traceback": traceback.format_exc(),
            "mode": "backtest"
        }


def process_single_subscriber_live(
    subscriber: Dict[str, Any],
    settings: Dict[str, Any],
    LiveTrader: type,
    df_with_signals: Any,
    idx: int,
    total: int
) -> Dict[str, Any]:
    """
    Process a single subscriber in parallel for live execution.

    Returns:
        {
            'user_id': str,
            'success': bool,
            'error': str (if failed)
        }
    """
    import logging

    user_id = subscriber.get('user_id')

    try:
        logging.info(f"\n   [{idx}/{total}] Processing user {user_id}...")

        # Create subscriber-specific settings
        subscriber_settings = {
            **settings,
            'user_id': subscriber['user_id'],  # For unique state file paths
            'api_key': subscriber['api_key'],
            'api_secret': subscriber['api_secret'],
            'leverage': subscriber.get('leverage', 10),
            'risk_per_trade': subscriber.get('risk_per_trade', 0.02),
            'initial_capital': subscriber.get('capital', 10000),
        }

        # Initialize LiveTrader for this subscriber
        subscriber_trader = LiveTrader(settings=subscriber_settings)

        # Execute their check_for_new_signal method
        subscriber_trader.check_for_new_signal(df_with_signals)

        logging.info(f"   ✅ User {user_id} processed successfully")

        return {
            'user_id': user_id,
            'success': True
        }

    except Exception as e:
        error_msg = str(e)
        logging.error(f"   ❌ User {user_id} failed: {error_msg}")
        logging.error(traceback.format_exc())

        return {
            'user_id': user_id,
            'success': False,
            'error': error_msg
        }


def execute_live(module, settings: Dict[str, Any], subscribers: List[Dict[str, Any]], strategy_dir: str) -> Dict[str, Any]:
    """
    Execute strategy in live mode with multiple subscribers using multi-tenant approach.

    Multi-Tenant Approach:
        1. Fetch market data ONCE (shared across all subscribers)
        2. Generate signals ONCE (same for all subscribers)
        3. Execute strategy for EACH subscriber with their credentials

    Args:
        module: Loaded strategy module
        settings: Strategy settings
        subscribers: List of subscriber configurations with API keys
        strategy_dir: Directory for logs and output

    Returns:
        Execution results as dictionary
    """
    import logging

    try:
        # Setup logging first
        CsvHandler = getattr(module, 'CsvHandler')
        logs_dir = os.path.join(strategy_dir, 'logs')
        os.makedirs(logs_dir, exist_ok=True)

        # Setup logging with strategy-specific log file
        log_filename = os.path.join(logs_dir, f"strategy_{settings.get('strategy_id', 'unknown')}.csv")
        CsvHandler.setup_logging(filename=log_filename)

        # Get required classes from module
        LiveTrader = getattr(module, 'LiveTrader')

        logging.info(f"🚀 Multi-Tenant Execution Started")
        logging.info(f"   Pair: {settings.get('pair')}")
        logging.info(f"   Resolution: {settings.get('resolution')}")
        logging.info(f"   Subscribers: {len(subscribers)}")

        # ====================================================================
        # STEP 1: Fetch market data ONCE (shared across all subscribers)
        # ====================================================================
        logging.info("📊 Fetching market data (shared across subscribers)...")

        # Use first subscriber's credentials just to fetch data
        first_subscriber = subscribers[0]
        temp_settings = {
            **settings,
            'api_key': first_subscriber['api_key'],
            'api_secret': first_subscriber['api_secret']
        }

        temp_trader = LiveTrader(settings=temp_settings)
        df = temp_trader.get_latest_data()

        if df is None or (hasattr(df, 'empty') and df.empty):
            raise ValueError("Failed to fetch market data")

        logging.info(f"   Fetched {len(df)} candles")

        # ====================================================================
        # STEP 2: Generate signals ONCE (same for all subscribers)
        # ====================================================================
        logging.info("🧠 Generating signals...")

        # Use the LiveTrader instance's inherited generate_signals method
        df_with_signals = temp_trader.generate_signals(df, settings)

        if df_with_signals is None or (hasattr(df_with_signals, 'empty') and df_with_signals.empty):
            raise ValueError("Signal generation returned empty dataframe")

        logging.info("✅ Signals generated")

        # ====================================================================
        # STEP 3: Execute strategy for EACH subscriber with their credentials (PARALLEL)
        # ====================================================================
        max_workers = settings.get('max_workers', 10)  # Configurable, default 10
        logging.info(f"💼 Processing {len(subscribers)} subscribers in parallel (max_workers={max_workers})...")

        subscribers_processed = 0
        trades_attempted = 0
        errors = {}

        # Process subscribers in parallel using ThreadPoolExecutor
        with ThreadPoolExecutor(max_workers=max_workers) as executor:
            # Submit all subscriber processing tasks
            future_to_subscriber = {
                executor.submit(
                    process_single_subscriber_live,
                    subscriber,
                    settings,
                    LiveTrader,
                    df_with_signals,
                    idx,
                    len(subscribers)
                ): subscriber
                for idx, subscriber in enumerate(subscribers, 1)
            }

            # Collect results as they complete
            for future in as_completed(future_to_subscriber):
                subscriber = future_to_subscriber[future]
                try:
                    result = future.result()

                    if result['success']:
                        subscribers_processed += 1
                        trades_attempted += 1
                    else:
                        errors[result['user_id']] = result.get('error', 'Unknown error')

                except Exception as e:
                    # This catches errors in the future itself (unlikely)
                    user_id = subscriber.get('user_id', 'unknown')
                    error_msg = str(e)
                    errors[user_id] = error_msg
                    logging.error(f"   ❌ Future exception for user {user_id}: {error_msg}")
                    logging.error(traceback.format_exc())

        # ====================================================================
        # STEP 4: Return results
        # ====================================================================
        logging.info(f"\n🏁 Multi-Tenant Execution Complete")
        logging.info(f"   Subscribers Processed: {subscribers_processed}/{len(subscribers)}")
        logging.info(f"   Trades Attempted: {trades_attempted}")
        logging.info(f"   Errors: {len(errors)}")

        return {
            "success": True,
            "mode": "live",
            "subscribers_processed": subscribers_processed,
            "trades_attempted": trades_attempted,
            "errors": errors if errors else None
        }

    except Exception as e:
        logging.error(f"💥 Multi-tenant execution failed: {e}")
        logging.error(traceback.format_exc())

        return {
            "success": False,
            "error": str(e),
            "traceback": traceback.format_exc(),
            "mode": "live",
            "subscribers_processed": 0,
            "trades_attempted": 0
        }


def execute_strategy(input_data: Dict[str, Any]) -> Dict[str, Any]:
    """
    Main execution function that routes to backtest or live mode

    Args:
        input_data: JSON input containing:
            - mode: "backtest" or "live"
            - strategy_file: absolute path to strategy .py file
            - settings: strategy configuration
            - subscribers: list of subscriber configs (for live mode only)

    Returns:
        Execution results as dictionary
    """
    try:
        # Validate input
        mode = input_data.get('mode')
        strategy_file = input_data.get('strategy_file')
        settings = input_data.get('settings', {})
        subscribers = input_data.get('subscribers', [])

        if not mode:
            return {
                "success": False,
                "error": "Missing 'mode' parameter"
            }

        if not strategy_file:
            return {
                "success": False,
                "error": "Missing 'strategy_file' parameter"
            }

        if mode not in ['backtest', 'live']:
            return {
                "success": False,
                "error": f"Invalid mode: {mode}. Must be 'backtest' or 'live'"
            }

        # Load the strategy module
        module = load_strategy_module(strategy_file)

        # Get strategy directory for logs
        strategy_dir = os.path.dirname(strategy_file)

        # Route to appropriate executor
        if mode == 'backtest':
            return execute_backtest(module, settings)
        else:  # mode == 'live'
            if not subscribers:
                return {
                    "success": False,
                    "error": "Live mode requires at least one subscriber",
                    "mode": "live"
                }
            return execute_live(module, settings, subscribers, strategy_dir)

    except Exception as e:
        return {
            "success": False,
            "error": str(e),
            "traceback": traceback.format_exc()
        }


def main():
    """
    Main entry point - reads JSON from stdin, executes strategy, outputs JSON to stdout
    """
    try:
        # Read input from stdin
        input_data = json.load(sys.stdin)

        # Execute strategy
        result = execute_strategy(input_data)

        # Output result as JSON (single line for parsing)
        print(json.dumps(result, default=str))

        # Exit with appropriate code
        sys.exit(0 if result.get('success') else 1)

    except json.JSONDecodeError as e:
        error_result = {
            "success": False,
            "error": f"Invalid JSON input: {str(e)}"
        }
        print(json.dumps(error_result))
        sys.exit(1)

    except Exception as e:
        error_result = {
            "success": False,
            "error": str(e),
            "traceback": traceback.format_exc()
        }
        print(json.dumps(error_result))
        sys.exit(1)


if __name__ == "__main__":
    main()
